### DISCUSSION TITLE
General Question

### DISCUSSION THREAD

**Discussion by itsmohammadh** [Cyfrin/foundry-full-course-cu#982] [Q&A]
Hi , i have a question i undrestand solidity code ( write and read code's and docs) but in lesson 9 i undrestand Raffle and Deploy and  another code but i confused in Test session and hard to undrestand how patrick write that( i mean logic for behind that ). What solution do you know for better understanding of test codes ?

**Comment by 0xusmanf** [comment]
What part of the tests you don't understand? Can you please be more specific?

> ( i mean logic for behind that )

What do you mean by this?

**Reply by itsmohammadh** [reply]
I mean the logic or thinking that is behind test writing, how to understand which part needs which test, I mean in general, I am not talking about a specific part.

**Reply by shikhar229169** [reply]
To be precise, I will say that test for all those cases for which the function call must be reverted. For example if in function you are saying that  if x1, x2, ...., xn conditions are satisfied that the function should revert. So you must cover all those in tests that with those test cases the function call is actually reverting.
And then write test for those cases for which the function should pass and do the required task. For example if you have a function which allows user to stake ETH in contract, then you should write a test in which first you stake the eth and test that the contract actually has the eth and also maintains correctly who deposited the eth.

**Comment by theirrationalone** [comment]
Take a deep breath first, then Please try to understand the concept. Make some notes on this, try to dig up semantically. If was in your place, First, I would focus on understanding the intention apart from coding stuff and will try to make some notes like this....

1. Okay we want a lottery, In which anybody can participate with a sufficient entry fee.
2. It's automatic lottery, It should announce the winner as time sufficient time passes on the way.....
3. How can i automate the lottery to announce the winner in a period of time.
4. How to write a code that can facilitate my automation.
5. Where that automation comes into???
6. Is it a independent On-Chain code logic or,
7. It's a Of-Chain mechanism??
8. Okay it's the Oracle chain link, I would like to understand the logic that how it facilitates or triggers time based logic into my code.
9. How should i react whenever my code logic announces a winner i.e.,
10. How should i update balances, number of participants, interval, lottery state, etc.
11. How should i clean up the lottery so it will be available again for participants, i.e., resetting timestamp.

Keep all these things in mind and try to write Test code on your own...

i can also give a example....

think on your own,

1. What should be in the lottery initially?? at very first lottery should a zero balance, zero participant, lottery should be open for anyone during the interval and initially.

2. If a participant entered into a lottery than, Lottery should update its own balance, number of players or participants from zero to 1.

3. If my lottery is calculating the winner, then nobody should be able to enter into the lottery during calculation.

4. Lottery must calculate the winner if enough time has been passed, lottery is in open state, lottery has some balance as well as associated players or participants.

Let's take the first case to write a test....

```solidity
function testLotteryHasZeroBalanceInitially() public {
    uint256 currentBalance = lottery.getBalance();
    
    assertEq(currentBalance, 0);
}
```

then after i got an error, In the error it is mentioned that `getBalance()` is not a function?? What does that mean???

It means that I haven't a getter function into my lottery called `getBalance()`. I can go back to my lottery and can create a getter function called `getBalance()`

at the bottom into lottery contract file....

```solidity
function getBalance() public view returns (uint256) {
    return address(this).balance;
}
```

or I have lottery itself. By using its address i can figure out its balance i.e.,...

```solidity
function testLotteryHasZeroBalanceInitially() public {
    uint256 currentBalance = address(lottery).balance.
    
    assertEq(currentBalance, 0);
}
```

Try to think deeply should happen when something happens.... and voila You will be able to write tests on your own. ;`)

**Reply by itsmohammadh** [reply]
thank you so much @theirrationalone
